src/qibo/models/error_mitigation.py

```diff
@@ -51,51 +51,77 @@ def get_gammas(noise_levels, analytical: bool = True):
     return zne_coefficients
 
 
-def get_noisy_circuit(circuit, num_insertions: int, insertion_gate: str = "CNOT"):
+def get_noisy_circuit(
+    circuit, num_insertions: int, global_unitary_folding=True, insertion_gate=None
+):
     """Standalone function to generate the noisy circuit with the inverse gate pairs insertions.
 
     Args:
         circuit (:class:`qibo.models.circuit.Circuit`): circuit to modify.
-        num_insertions (int): number of insertion gate pairs to add.
+        num_insertions (int): number of insertion gate pairs / global unitary folds to add.
+        global_unitary_folding (bool): Default True. Global unitary folding is carried out by default.
+            Global unitary folding takes precedence over insertion_gate.
         insertion_gate (str, optional): gate to be used in the insertion.
             If ``"RX"``, the gate used is :math:``RX(\\pi / 2)``.
-            Default is ``"CNOT"``.
+            Otherwise, the other gate is ``"CNOT"``.
 
     Returns:
-        :class:`qibo.models.Circuit`: circuit with the inserted gate pairs.
+        :class:`qibo.models.Circuit`: circuit with the inserted gate pairs or with global folding.
     """
-    if insertion_gate not in ("CNOT", "RX"):  # pragma: no cover
-        raise_error(
-            ValueError,
-            "Invalid insertion gate specification. Please select between 'CNOT' and 'RX'.",
-        )
-    if insertion_gate == "CNOT" and circuit.nqubits < 2:  # pragma: no cover
-        raise_error(
-            ValueError,
-            "Provide a circuit with at least 2 qubits when using the 'CNOT' insertion gate. "
-            + "Alternatively, try with the 'RX' insertion gate instead.",
-        )
 
-    i_gate = gates.CNOT if insertion_gate == "CNOT" else gates.RX
+    if global_unitary_folding:  # pragma: no cover
+        circuit_no_meas = circuit.__class__(**circuit.init_kwargs)
+        circuit_meas = circuit.__class__(**circuit.init_kwargs)
```

Comment:
Are these copies of the circuit? If yes, I would use the `circuit.copy` method.

Reply:
Hi Matteo, the purpose of lines L73 to L79, which reads,
```python
circuit_no_meas = circuit.__class__(**circuit.init_kwargs)
circuit_meas = circuit.__class__(**circuit.init_kwargs)
for gate in circuit.queue:
    if gate.name != "measure":
        circuit_no_meas.add(gate)
    else:
        circuit_meas.add(gate)
```
is to duplicate the input `circuit`. As its name suggests, `circuit_no_meas` is the input `circuit` without the measurements. Likewise, `circuit_meas` is the input `circuit` with measurements.

The reason I opted to have lines L73 to L79 is so that we can then construct our `noisy_circuit` by first creating a copy of `circuit_no_meas`.

Then we perform global unitary folding in lines L83 to L86, which reads,
```python
for _ in range(num_insertions):
    noisy_circuit += circuit_no_meas.invert() + circuit_no_meas

    noisy_circuit += circuit_meas
```

Would you advise to use `circuit.copy` or `Circuit(**circuit.init_kwargs)` to copy the circuit? What does the latter do?

Thanks very much!

Reply:
What I mean is that you don't need two copies of the circuit to do what you aim to do:

```python
from qibo import Circuit, gates

# build your circuit
c = Circuit(4)
for q in range(4):
    c.add(gates.H(q))
c.add(gates.CNOT(0,1))
c.add(gates.CNOT(3,1))
c.add(gates.M(0,3))

# initialize a copy with same arguments (dense, accelerators, etc)
copy_c = Circuit(**c.init_kwargs)

# populate the copy with non-M gates
for g in c.queue:
    if not isinstance(g, gates.M):
        copy_c.add(g)

# start building noisy circuit with U^+ and U
noisy_c = copy_c.invert() + copy_c

# add measurements according to original circuit
for m in c.measurements:
    noisy_c.add(m)
```

Reply:
Right! Thanks! I will use your implementation in the code. But we will need to add one more line before we build the noisy circuit so that our final circuit is `U (U^+ U)^n` for `n` repetitions of `(U^+ U)`.

```python
# initialize `noisy_c` as `copy_c`.
noisy_c = copy_c

# start building noisy circuit with U^+ and U
noisy_c = copy_c.invert() + copy_c

# add measurements according to original circuit
for m in c.measurements:
    noisy_c.add(m)
```